### PR TITLE
Default to stack.yml when no --yaml flag given

### PR DESCRIPTION
**Pull request by nicholasjackson**
## Description
Added capability to load a default configuration file `stack.yaml` when present in the current folder.  If a default is present this can still be overridden by providing the flag `--yaml=file.yaml`

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?
Code has been tested on a unit level and basic manual testing has also been performed.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -71,7 +71,7 @@ Specify `lang: Dockerfile` if you want the faas-cli to execute a build or `skip_
 
 A YAML stack file groups functions together and also saves on typing.
 
-You can define individual functions or a set of of them within a YAML file. This makes the CLI easier to use and means you can use this file to deploy to your OpenFaaS instance.
+You can define individual functions or a set of of them within a YAML file. This makes the CLI easier to use and means you can use this file to deploy to your OpenFaaS instance.  By default the faas-cli tool will attempt to load `stack.yaml` from the current directory.
```
Comment (alexellis) [inline README.md]: `faas-cli`

File: commands/faas_test.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+	if yamlFile != "stack.yml" {
+		t.Fatal(fmt.Printf("Expected yamlFile to equal %v got %v\n", "stack.yml", yamlFile))
+	}
+}
+
+func TestLoadsFromPametersYAMLWhenGivenAndDefaultPresent(t *testing.T) {
+	rs := resetState()
+	defer rs()
+	os.Chdir("./test_fixtures")
```
Comment (ericstoekl) [inline commands/faas_test.go]: Would you be okay with changing the name of the directory to `testdata`? We have a [PR out](https://github.com/openfaas/faas-cli/pull/87) that creates this directory. Also, the go source code mentions that `testdata` dirs are [ignored by the compiler](https://go-review.googlesource.com/c/go/+/23353/1/src/cmd/go/test.go#62). Let me know your thoughts. Thanks!

File: commands/faas_test.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+	Execute([]string{"help"})
+
+	if yamlFile != "stack.yml" {
+		t.Fatal(fmt.Printf("Expected yamlFile to equal %v got %v\n", "stack.yml", yamlFile))
+	}
+}
+
+func TestLoadsFromPametersYAMLWhenGivenAndDefaultPresent(t *testing.T) {
```
Comment (ericstoekl) [inline commands/faas_test.go]: `..PametersYAML..`
Is there a typo here?

File: commands/faas.go [hunk 4]
```diff
@@ -5,12 +5,15 @@ package commands
...
 	"fmt"
+	"os"
 
 	"github.com/spf13/cobra"
 )
 
 const defaultGateway = "http://localhost:8080"
+
 const defaultNetwork = "func_functions"
+const defaultYaml = "stack.yml"
```
Comment (ericstoekl) [inline commands/faas.go]: Would you be open to having two const variables here:

`defaultYml = "stack.yml"`
`defaultYaml = "stack.yaml"`

Then have a `else if ` block for `defaultYml` on line 59

File: commands/faas_test.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	os.Chdir("./test_fixtures")
+
+	Execute([]string{"help", "--yaml=myfile.yml"})
+
+	if yamlFile != "myfile.yml" {
+		t.Fatal(fmt.Printf("Expected yamlFile to equal %v got %v\n", "stack.yml", yamlFile))
+	}
+}
+
+func TestDoesNotLoadsDefaultYAMLWhenMissing(t *testing.T) {
```
Comment (ericstoekl) [inline commands/faas_test.go]: `...LoadDefaultYAML....`

File: commands/faas_test.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+func TestLoadsFromPametersYAMLWhenGivenAndDefaultPresent(t *testing.T) {
+	rs := resetState()
+	defer rs()
+	os.Chdir("./test_fixtures")
+
+	Execute([]string{"help", "--yaml=myfile.yml"})
+
+	if yamlFile != "myfile.yml" {
+		t.Fatal(fmt.Printf("Expected yamlFile to equal %v got %v\n", "stack.yml", yamlFile))
```
Comment (ericstoekl) [inline commands/faas_test.go]: Why not `t.Fatalf("...")` instead?

File: commands/faas.go (hunk 4 above)
Comment (alexellis) [reply commands/faas.go]: Hmm.. I think we can assume `yml`only for a default vs `yaml`. Having had a quick google "YML" as an extension is the winner for usage.

File: commands/faas_test.go (hunk 6 above)
Comment (nicholasjackson) [reply commands/faas_test.go]: Good catch, TBH, I never use the the default testing framework in go, I find you have to write way too much boilerplate.  I generally use this https://github.com/stretchr/testify

Comment (alexellis) [reply commands/faas_test.go]: Thanks for matching the project style 💯

File: commands/faas.go [hunk 7]
```diff
@@ -5,12 +5,15 @@ package commands
 
 import (
 	"fmt"
+	"os"
 
 	"github.com/spf13/cobra"
 )
 
 const defaultGateway = "http://localhost:8080"
+
```
Comment (itscaro) [inline commands/faas.go]: Is it possible to remove this blank line and use const() instead?

**Comment by nicholasjackson** [comment]
I think all the requested changes have been completed, I have also signed off the commit hopefully all good

File: commands/faas_test.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+)
+
+func TestLoadsDefaultYAMLWhenPresent(t *testing.T) {
+	rs := resetState()
+	defer rs()
+	os.Chdir("./testdata")
```
Comment (alexellis) [inline commands/faas_test.go]: I think this works, but wondered if we could fake the result of `os.Stat`?

Comment (alexellis) [reply commands/faas_test.go]: In this usage I effectively "faked" OS env:

https://github.com/openfaas/faas/blob/master/watchdog/config_test.go#L34

**Comment by alexellis** [comment]
Good to get the test with this 👍 I added a comment around just faking the os.Stat and returning a canned value.

**Comment by alexellis** [comment]
Trying to test this out locally.. what am I getting wrong?

```
alexr:faas-cli alex$ ./faas-cli build
Please provide a valid -image name for your Docker image.
alexr:faas-cli alex$ git branch
* default-yaml
  master
  proxy
  proxy-test
alexr:faas-cli alex$ ls ./stack.yml
./stack.yml
alexr:faas-cli alex$ ./faas-cli build
Please provide a valid -image name for your Docker image.
alexr:faas-cli alex$ ./faas-cli deploy
Please provide a --image to be deployed.
```

**Comment by alexellis** [comment]
This tests out OK for me. Now that we're building an experience around a default .yml file - should we think about changing the `new` functionality to write/append to "stack.yml" by default? cc/ @ericstoekl @rgee0 @johnmccabe

**Comment by rgee0** [comment]
@alexellis I was thinking along those lines after reviewing @ericstoekl's work on the append functionality.  I think it would make sense, and result in a better experience to default it and allow over-rides.

**Comment by alexellis** [comment]
@nicholasjackson is this good to go? Anything else needed?

**Comment by alexellis** [comment]
@itscaro can you test?

File: commands/faas.go [hunk 9]
```diff
@@ -42,11 +50,20 @@ func init() {
...
 // Execute TODO
 func Execute(customArgs []string) {
+	checkAndSetDefaultYaml()
+
 	faasCmd.SilenceUsage = true
 	faasCmd.SetArgs(customArgs[1:])
 	faasCmd.Execute()
 }
 
+func checkAndSetDefaultYaml() {
```
Comment (alexellis) [inline commands/faas.go]: I want to get the fix in, think this is going to be a good addition. This method feels hacky to do this to a global variable. @johnmccabe any other suggestions?

**Pull request merged**